cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java

```diff
@@ -0,0 +1,229 @@
+      throw new IOException(String.format("Parent Directory for %s is null", logFile.toURI().toString()));
+    }
+
+    // check if parent directories exist
+    mkdirsIfNotExists(logDir, dirPermissions);
+
+    LOG.info("Created log directory: {}", logDir.toURI().toString());
+    if (logFile.exists()) {
+      // The file name for a given application exists if the appender was stopped and then started again but file was
+      // not rolled over. In this case, since the roll over size is typically small, we can rename the old file and
+      // copy its contents to new file and delete old file.
+      long now = System.currentTimeMillis();
+
+      // rename existing file to temp file
+      Location tempLocation = logFile.renameTo(logDir.append("temp-" + Long.toString(now)));
+
+      if (tempLocation == null) {
+        throw new IOException(String.format("Can not rename file %s", logFile.toURI().toString()));
+      }
+
+      try (InputStream inputStream = tempLocation.getInputStream()) {
+        // create new file and open outputstream on it
+        logFile.createNew(filePermissions);
+        // TODO: Handle existing file in a better way rather than copying it over
+        OutputStream outputStream = new LocationOutputStream(logFile, logFile.getOutputStream());
+        activeLocations.put(locationIdentifier, (LocationOutputStream) outputStream);
+        ByteStreams.copy(inputStream, outputStream);
+      } catch (IOException e) {
+        activeLocations.remove(locationIdentifier);
+        throw e;
+      }
+
+
+      try {
+        // clean up all the temp files for which deletion failed
+        tempLocation.delete();
+        for (Location location : logDir.list()) {
+          if (location.toURI().toString().contains("temp-")) {
+            location.delete();
+          }
+        }
+      } catch (IOException e) {
+        // do not throw any exception while deleting temp directory
+        LOG.warn("Not able to delete temp location, will be retried later");
+      }
+
+    } else {
+      // create file with correct permissions
+      logFile.createNew(filePermissions);
+      activeLocations.put(locationIdentifier, new LocationOutputStream(logFile, logFile.getOutputStream()));
+    }
+
+    return activeLocations.get(locationIdentifier);
+  }
+
+  /**
+   * Closes all open output streams and clears cache
+   */
+  public void close() {
+    Collection<LocationOutputStream> locations = new ArrayList<>(activeLocations.values());
+    activeLocations.clear();
+
+    for (LocationOutputStream locationOutputStream : locations) {
+      // we do not want to throw any exception rather close all the open output streams. so close quietly
+      Closeables.closeQuietly(locationOutputStream);
+    }
+  }
+
+  /**
+   * Flushes all the open output streams
+   */
+  @Override
+  public void flush() throws IOException {
+    Collection<LocationOutputStream> locations = new ArrayList<>(activeLocations.values());
```

Comment:
you don't have to make a copy of values for flush. for close we make a copy to avoid concurrentModificationException, as when we close the stream, we are remove the elements from the map of active locations.